File: src/docforge/openapi/iter.py
from collections.abc import Iterator
from dataclasses import dataclass, field
from typing import Any

HTTP_METHODS = {"get", "post", "put", "delete", "patch", "head", "options", "trace"}


@dataclass(frozen=True)
class Endpoint:
    method: str
    path: str
    operation: dict[str, Any]
    tags: list[str] = field(default_factory=list)
    summary: str = ""
    description: str = ""


@dataclass(frozen=True)
class Schema:
    name: str
    body: dict[str, Any]


def iter_endpoints(spec: dict[str, Any]) -> Iterator[Endpoint]:
    paths = spec.get("paths") or {}
    for path, item in paths.items():
        if not isinstance(item, dict):
            continue
        for method, op in item.items():
            if method.lower() not in HTTP_METHODS:
                continue
            if not isinstance(op, dict):
                continue
            yield Endpoint(
                method=method.lower(),
                path=path,
                operation=op,
                tags=list(op.get("tags") or []),
                summary=str(op.get("summary") or ""),
                description=str(op.get("description") or ""),
            )


def iter_schemas(spec: dict[str, Any]) -> Iterator[Schema]:
    schemas = (spec.get("components") or {}).get("schemas") or {}
    for name, body in schemas.items():
        if not isinstance(body, dict):
            continue
        yield Schema(name=name, body=body)

File: src/openapi/iter.ts
const HTTP_METHODS = new Set([
  "get", "post", "put", "delete", "patch", "head", "options", "trace",
]);

export interface Endpoint {
  method: string;
  path: string;
  operation: Record<string, unknown>;
  tags: string[];
  summary: string;
  description: string;
}

export interface Schema {
  name: string;
  body: Record<string, unknown>;
}

function isPlainObject(x: unknown): x is Record<string, unknown> {
  return x !== null && typeof x === "object" && !Array.isArray(x);
}

export function* iterEndpoints(
  spec: Record<string, unknown>,
): Generator<Endpoint> {
  const paths = isPlainObject(spec.paths) ? spec.paths : {};
  for (const [path, item] of Object.entries(paths)) {
    if (!isPlainObject(item)) continue;
    for (const [method, op] of Object.entries(item)) {
      const lower = method.toLowerCase();
      if (!HTTP_METHODS.has(lower)) continue;
      if (!isPlainObject(op)) continue;
      const tagsRaw = op.tags;
      const tags = Array.isArray(tagsRaw)
        ? tagsRaw.map((t) => String(t))
        : [];
      yield {
        method: lower,
        path,
        operation: op,
        tags,
        summary: typeof op.summary === "string" ? op.summary : "",
        description: typeof op.description === "string" ? op.description : "",
      };
    }
  }
}

export function* iterSchemas(
  spec: Record<string, unknown>,
): Generator<Schema> {
  const components = isPlainObject(spec.components) ? spec.components : {};
  const schemas = isPlainObject(components.schemas) ? components.schemas : {};
  for (const [name, body] of Object.entries(schemas)) {
    if (!isPlainObject(body)) continue;
    yield { name, body };
  }
}
